src/config.rs

```diff
@@ -0,0 +1,150 @@
+use anyhow::{bail, Result};
+use std::path::{Path, PathBuf};
+
+#[cfg_attr(feature = "serde", derive(serde::Deserialize))]
+#[derive(Clone, PartialEq, Eq, Debug)]
+pub struct Config {
+    #[cfg_attr(feature = "serde", serde(default))]
+    pub target: Target,
+    #[cfg_attr(feature = "serde", serde(default))]
+    pub atomics: bool,
+    #[cfg_attr(feature = "serde", serde(default))]
+    pub atomics_feature: Option<String>,
+    #[cfg_attr(feature = "serde", serde(default))]
+    pub generic_mod: bool,
+    #[cfg_attr(feature = "serde", serde(default))]
+    pub make_mod: bool,
+    #[cfg_attr(feature = "serde", serde(default))]
+    pub ignore_groups: bool,
+    #[cfg_attr(feature = "serde", serde(default))]
+    pub keep_list: bool,
+    #[cfg_attr(feature = "serde", serde(default))]
+    pub strict: bool,
+    #[cfg_attr(feature = "serde", serde(default))]
+    pub pascal_enum_values: bool,
+    #[cfg_attr(feature = "serde", serde(default))]
+    pub feature_group: bool,
+    #[cfg_attr(feature = "serde", serde(default))]
+    pub feature_peripheral: bool,
+    #[cfg_attr(feature = "serde", serde(default))]
+    pub max_cluster_size: bool,
+    #[cfg_attr(feature = "serde", serde(default))]
+    pub impl_debug: bool,
+    #[cfg_attr(feature = "serde", serde(default))]
+    pub impl_debug_feature: Option<String>,
+    #[cfg_attr(feature = "serde", serde(default = "current_dir"))]
+    pub output_dir: PathBuf,
+    #[cfg_attr(feature = "serde", serde(default))]
+    pub input: Option<PathBuf>,
+    #[cfg_attr(feature = "serde", serde(default))]
+    pub source_type: SourceType,
+    #[cfg_attr(feature = "serde", serde(default))]
+    pub log_level: Option<String>,
+    #[cfg_attr(feature = "serde", serde(default))]
+    pub interrupt_link_section: Option<String>,
+    #[cfg_attr(feature = "serde", serde(default))]
+    pub reexport_core_peripherals: bool,
+    #[cfg_attr(feature = "serde", serde(default))]
+    pub reexport_interrupt: bool,
+}
```

Comment:
all of these `#[cfg_attr(feature = "serde", serde(default))]` (and the one with the function call) could be replaced with a `#[cfg_attr(feature = "serde", serde(default))]` on the container I think

https://serde.rs/container-attrs.html#default

```suggestion
#[cfg_attr(feature = "serde", derive(serde::Deserialize))]
#[derive(Clone, PartialEq, Eq, Debug)]
#[cfg_attr(feature = "serde", serde(default))]
pub struct Config {
    pub target: Target,
    pub atomics: bool,
    pub atomics_feature: Option<String>,
    pub generic_mod: bool,
    pub make_mod: bool,
    pub ignore_groups: bool,
    pub keep_list: bool,
    pub strict: bool,
    pub pascal_enum_values: bool,
    pub feature_group: bool,
    pub feature_peripheral: bool,
    pub max_cluster_size: bool,
    pub impl_debug: bool,
    pub impl_debug_feature: Option<String>,
    pub output_dir: PathBuf,
    pub input: Option<PathBuf>,
    pub source_type: SourceType,
    pub log_level: Option<String>,
    pub interrupt_link_section: Option<String>,
    pub reexport_core_peripherals: bool,
    pub reexport_interrupt: bool,
}
```

Reply:
this would also bring another benefit, because I think it expands to

```rs
Config {
    <fields that were deserialized>,
    ..Config::default()
}
```
meaning the deserialization of json `{}` = `Config::default()` always.

Reply:
What about `output_dir`?

Reply:
it should be fine, since it'd use the current_dir call in `<Config as Default>::default()`

Reply:
I've changed `output_dir` to `Option<PathBuf>`.